File: temp/arena.py

# Binary Search on Answer

def miniMax(nums, low, high):

    def isMidValid(mid):
        pass

    ans = high
    while low <= high:

        mid = (low + high) // 2

        if isMidValid(mid):
            ans = mid
            high = mid - 1
        else:
            low = mid + 1

    return ans

def maxiMin(nums, low, high):

    def isMidValid(mid):
        pass

    ans = low
    while low <= high:

        mid = (low + high) // 2

        if isMidValid(mid):
            ans = mid
            low = mid + 1
        else:
            high = mid - 1

    return ans

# Trees

def preorder(root):
    if root == None:
        return []
    return [root.val] + preorder(root.left) + preorder(root.right)

def inorder(root):
    if root == None:
        return []
    return inorder(root.left) + [root.val] + inorder(root.right)

def postorder(root):
    if root == None:
        return []
    return preorder(root.left) + preorder(root.right) + [root.val]

def morris(root):

    res = []

    while root != None:

        if root.left == None:
            res.append(root.val)
            root = root.right
        else:

            predec = root.left
            while predec.right and predec.right != root:
                predec = predec.right

            if predec.right == None:
                predec.right = root
                root = root.left
            else:
                predec.right = None
                res.append(root.val)
                root = root.right
    return res

from collections import deque
from math import *

def bfs(root):

    res = []
    queue = deque([root])

    while queue:

        n = len(queue)
        level = []

        for _ in range(n):

            node = queue.popleft()
            level.append(node.val)

            for nei in [node.left, node.right]:
                if nei:
                    queue.append(nei)

        res.append(level)

    return res

class BIT:

    def __init__(self, n, nums):
        self.bit = [0] * (n + 1)
        self.n = n

        for idx in range(n):
            self.update(idx, nums[idx])

    def update(self, idx, delta):

        idx += 1

        while idx <= self.n:
            self.bit[idx] += delta
            idx += idx & -idx

    def query(self, idx):

        idx += 1
        res = 0

        while idx > 0:
            res += self.bit[idx]
            idx -= idx & -idx

        return res

class SegmentTree:

    def __init__(self, nums):

        self.n = len(nums)
        self.seg = [0] * (4 * self.n)
        self.nums = nums
        self.build(1, 0, self.n - 1)

    def combine(self, left, right):
        return left + right
    
    def build(self, idx, left, right):

        if left == right:
            self.seg[idx] = self.nums[left]
            return
        
        mid = (left + right) // 2
        self.build(2 * idx, left, mid)
        self.build(2 * idx + 1, mid + 1, right)
        self.seg[idx] = self.combine(self.seg[2 * idx], self.seg[2 * idx + 1])

    def query(self, idx, left, right, ql, qr):
        
        if ql > right or qr < left:
            return  0
        
        if ql <= left and right <= qr:
            return self.seg[idx]
        
        mid = (left + right) // 2

        return self.combine(
            self.query(2 * idx, left, mid, ql, qr),
            self.query(2 * idx + 1, mid + 1, right, ql, qr)
        )
    
    def update(self, idx, left, right, pos, val):

        if left == right:
            self.seg[idx] = val
            return
        
        mid = (left + right) // 2

        if pos <= mid:
            self.update(2 * idx, left, mid, pos, val)
        else:
            self.update(2 * idx + 1, mid + 1, right, pos, val)

        self.seg[idx] = self.combine(self.seg[2 * idx], self.seg[2 * idx + 1])

class STLP:

    def __init__(self, nums):
        
        self.n = len(nums)
        self.nums = nums
        self.seg = [0] * (4 * self.n)
        self.lazy = [0] * (4 * self.n)

        self.build(1, 0, self.n - 1)

    def combine(self, left, right):
        return left + right

    def build(self, idx, left, right):

        if left == right:
            self.seg[idx] = self.nums[left]
            return
        
        mid = (left + right) // 2

        self.build(2 * idx, left, mid)
        self.build(2 * idx + 1, mid + 1, right)

        self.seg[idx] = self.combine(self.seg[2 * idx], self.seg[2 * idx + 1])

    def propagate(self, idx, left, right):

        if self.lazy[idx] != 0:

            self.seg[idx] += (right - left + 1) * self.lazy[idx]

            if left != right:
                self.lazy[2 * idx] += self.lazy[idx]
                self.lazy[2 * idx + 1] += self.lazy[idx]

            self.lazy[idx] = 0
    
    def query(self, idx, left, right, ql, qr):

        self.propagate(idx, left, right)

        if ql > right or qr < left:
            return 0
        
        if ql <= left and right <= qr:
            return self.seg[idx]
        
        mid = (left + right) // 2

        return self.combine(
            self.query(2 * idx, left, mid, ql, qr),
            self.query(2 * idx + 1, mid + 1, right, ql, qr)
        )
    
    def update(self, idx, left, right, ql, qr, val):

        self.propagate(idx, left, right)

        if ql > right or qr < left:
            return
        
        if ql <= left and right <= qr:
            self.lazy[idx] += val
            self.propagate(idx, left, right)
            return
        
        mid = (left + right) // 2

        self.update(2 * idx, left, mid, ql, qr, val)
        self.update(2 * idx + 1, mid + 1, right, ql, qr, val)

        self.seg[idx] = self.combine(self.seg[2 * idx], self.seg[2 * idx + 1])

    def range_query(self, left, right):
        return self.query(1, 0, self.n - 1, left, right)
    
    def range_update(self, left, right, val):
        self.update(1, 0, self.n - 1, left, right, val)

# Graphs

def dfs(graph, node, seen = None):

    if seen == None:
        seen = set()

    res = [node]
    seen.add(node)

    for nei in graph[node]:
        if nei not in seen:
            res.extend(dfs(graph, nei, seen))

    return res

def bfs(root, graph):

    res = []
    queue = deque([root])
    seen = set([root])

    while queue:
        node = queue.popleft()
        res.append(node)

        for nei in graph[node]:
            if nei not in seen:
                queue.append(nei)
                seen.add(nei)

    return res

from collections import defaultdict

def topoSort(edges, n, oneBased=0):

    graph = defaultdict(list)
    indegree = [0] * (n + oneBased)

    for u, v in edges:
        graph[u].append(v)
        indegree[v] += 1

    topo = []

    queue = deque([node for node in range(oneBased, n + oneBased) if indegree[node] == 0])

    while queue:
        node = queue.popleft()
        topo.append(node)

        for nei in graph[node]:
                
            indegree[nei] -= 1

            if indegree[nei] == 0:
                queue.append(nei)

    if len(topo) != n:
        return -1

    return topo

def isBipartite(graph, n):

    color = [-1] * n

    for start in range(n):

        if color[start] == -1:

            queue = deque([start])
            color[start] = 1

            while queue:

                node = queue.popleft()

                for nei in graph[node]:

                    if color[nei] == -1:
                        color[nei] = 1 - color[node]
                    elif color[nei] == color[node]:
                        return False

    return True

import heapq

def dijkstra(source, graph, n):

    dist = [float("inf")] * n

    queue = [(0, source)]
    dist[source] = 0

    while queue:

        d, u = heapq.heappop(queue)

        if d > dist[u]:
            continue

        for v, w in graph[u]:

            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                heapq.heappush(queue, (dist[v], v))

    return dist

def bellmanFord(source, graph, n):

    dist = [float("inf")] * n
    dist[source] = 0

    for _ in range(n - 1):
        for u in range(n):
            for v, w in graph[u]:
                if dist[u] != float("inf") and dist[u] + w < dist[v]:
                    dist[v] = dist[u] + w

    for u in range(n):
        for v, w in graph[u]:
            if dist[u] != float("inf") and dist[u] + w < dist[v]:
                return -1
            
    return dist

def floydWarshall(edges, n):

    dist = [[float("inf")] * n for _ in range(n)]

    for i in range(n):
        dist[i][i] = 0

    for u, v, w in edges:
        dist[u][v] = min(dist[u][v], w)

    for k in range(n):
        for u in range(n):
            for v in range(n):
                if dist[u][k] != float("inf")  and dist[k][v] != float("inf"):
                    dist[u][v] = min(dist[u][v], dist[u][k] + dist[k][v])

    for i in range(n):
        if dist[i][i] < 0:
            return -1
        
    return dist

class DSU:

    def __init__(self, n):
        self.rank = [0] * n
        self.size = [1] * n
        self.parent = [node for node in range(n)]
        self.components = n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y):
        px, py = self.find(x), self.find(y)

        if px == py:
            return False
        
        if self.rank[py] < self.rank[px]:
            self.parent[py] = px
        elif self.rank[px] < self.rank[py]:
            self.parent[px] = py
        else:
            self.parent[py] = px
            self.rank[px] += 1
        
        self.components -= 1
        return True
    
def kruskal(edges, n):

    dsu = DSU(n)
    mst = []
    weight = 0

    minHeap = [(w, u, v) for u, v, w in edges]
    heapq.heapify(minHeap)

    while minHeap and len(mst) < n - 1:

        w, u, v = heapq.heappop(minHeap)

        if dsu.union(u, v):
            mst.append((u, v, w))
            weight += w

    if len(mst) != n - 1:
        return -1

    return mst, weight

def prims(graph, start = 0):

    mst = []
    seen = set()
    totalWeight = 0

    pq = [(0, start, -1)]

    while pq:
        weight, node, parent = heapq.heappop(pq)

        if node in seen:
            continue

        seen.add(node)

        if parent != -1:
            mst.append((parent, node, weight))
            totalWeight += weight

        for v, w in graph[node]:
            if v not in seen:
                heapq.heappush(pq, (w, v, node))

    return mst, totalWeight

def kosaraju(graph, n):
    
    seen = set()
    components = []
    finishingOrder = []

    def dfs(node):

        seen.add(node)

        for nei in graph[node]:
            if nei not in seen:
                dfs(nei)

        finishingOrder.append(node)

    for node in range(n):
        if node not in seen:
            dfs(node)

    transposed = defaultdict(list)

    for u in range(n):
        for v in graph[u]:
            transposed[v].append(u)

    seen.clear()

    def dfs2(node, comp):
        seen.add(node)
        comp.append(node)

        for nei in transposed[node]:
            if nei not in seen:
                dfs2(nei, comp)
    
    for node in reversed(finishingOrder):

        if node not in seen:
            comp = []
            dfs2(node, comp)
            components.append(comp)
    
    return components, len(components)

def tarjan(graph, n):

    def dfs(node, parent, visited, graph, tin, low, bridges, timer):

        visited.add(node)
        tin[node] = low[node] = timer
        timer[0] += 1

        for nei in graph[node]:

            if nei == parent:
                continue

            if nei not in visited:
                dfs(nei, node, visited, graph, tin, low, bridges, timer)
                low[node] = min(low[node], low[nei])

                if low[nei] > tin[node]:
                    bridges.append((node, nei))
            else:
                low[node] = min(low[node], tin[nei])


    bridges = []

    tin = [0] * n
    low = [0] * n
    timer = [1]
    visited = set()

    for i in range(n):
        if i not in visited:
            dfs(i, -1, visited, graph, tin, low, bridges, timer)

    return bridges

def articulationPoints(graph, n):

    def dfs(node, parent, graph, visited, low, tin, timer, ans):

        visited.add(node)
        tin[node] = low[node] = timer[0]
        timer[0] += 1
        child = 0

        for nei in graph[node]:

            if nei == parent:
                continue

            if nei not in visited:
                dfs(nei, node, graph, visited, low, tin, timer, ans)

                low[node] = min(low[node], low[nei])

                if low[nei] >= tin[node] and parent != -1:
                    ans.add(node)
                
                child += 1
            else:
                low[node] = min(low[node], tin[nei])

        if child > 1 and parent == -1:
            ans.add(node)

    ans = set()
    visited = set()
    low = [0] * n
    tin = [0] * n
    timer = [0]

    for node in range(n):
        if node not in visited:
            dfs(node, -1, graph, visited, low, tin, timer, ans)

    return list(ans) if ans else [-1]

File: arena.cpp
#include <bits/stdc++.h>
using namespace std;

// Binary Search
int binarySearch(vector<int>& arr, int target) {
    int low = 0, high = arr.size() - 1;
    
    while (low <= high) {
        int mid = low + (high - low) / 2;
        
        if (arr[mid] == target) return mid;
        else if (arr[mid] < target) low = mid + 1;
        else high = mid - 1;
    }
    
    return -1;
}

int binarySearchAnswer(int low, int high, function<bool(int)> isMidValid) {
    int ans = -1;
    
    while (low <= high) {
        int mid = low + (high - low) / 2;
        
        if (isMidValid(mid)) {
            ans = mid;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    
    return ans;
}

// Tree Node Structure
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

// Tree Traversals
void preorder(TreeNode* root, vector<int>& res) {
    if (!root) return;
    res.push_back(root->val);
    preorder(root->left, res);
    preorder(root->right, res);
}

void inorder(TreeNode* root, vector<int>& res) {
    if (!root) return;
    inorder(root->left, res);
    res.push_back(root->val);
    inorder(root->right, res);
}

void postorder(TreeNode* root, vector<int>& res) {
    if (!root) return;
    postorder(root->left, res);
    postorder(root->right, res);
    res.push_back(root->val);
}

vector<int> morris(TreeNode* root) {
    vector<int> res;
    
    while (root) {
        if (!root->left) {
            res.push_back(root->val);
            root = root->right;
        } else {
            TreeNode* pred = root->left;
            while (pred->right && pred->right != root) {
                pred = pred->right;
            }
            
            if (!pred->right) {
                pred->right = root;
                root = root->left;
            } else {
                pred->right = nullptr;
                res.push_back(root->val);
                root = root->right;
            }
        }
    }
    
    return res;
}

vector<vector<int>> bfsTree(TreeNode* root) {
    vector<vector<int>> res;
    if (!root) return res;
    
    queue<TreeNode*> q;
    q.push(root);
    
    while (!q.empty()) {
        int n = q.size();
        vector<int> level;
        
        for (int i = 0; i < n; i++) {
            TreeNode* node = q.front();
            q.pop();
            level.push_back(node->val);
            
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
        
        res.push_back(level);
    }
    
    return res;
}

// Binary Indexed Tree
class BIT {
public:
    vector<int> bit;
    int n;
    
    BIT(int size, vector<int>& nums) : n(size) {
        bit.assign(n + 1, 0);
        for (int i = 0; i < n; i++) {
            update(i, nums[i]);
        }
    }
    
    void update(int idx, int delta) {
        idx++;
        while (idx <= n) {
            bit[idx] += delta;
            idx += idx & -idx;
        }
    }
    
    int query(int idx) {
        idx++;
        int res = 0;
        while (idx > 0) {
            res += bit[idx];
            idx -= idx & -idx;
        }
        return res;
    }
};

// Segment Tree
class SegmentTree {
public:
    vector<int> seg, nums;
    int n;
    
    SegmentTree(vector<int>& arr) : nums(arr), n(arr.size()) {
        seg.assign(4 * n, 0);
        build(1, 0, n - 1);
    }
    
    int combine(int left, int right) {
        return left + right;
    }
    
    void build(int idx, int left, int right) {
        if (left == right) {
            seg[idx] = nums[left];
            return;
        }
        
        int mid = (left + right) / 2;
        build(2 * idx, left, mid);
        build(2 * idx + 1, mid + 1, right);
        seg[idx] = combine(seg[2 * idx], seg[2 * idx + 1]);
    }
    
    int query(int idx, int left, int right, int ql, int qr) {
        if (ql > right || qr < left) return 0;
        if (ql <= left && right <= qr) return seg[idx];
        
        int mid = (left + right) / 2;
        return combine(
            query(2 * idx, left, mid, ql, qr),
            query(2 * idx + 1, mid + 1, right, ql, qr)
        );
    }
    
    void update(int idx, int left, int right, int pos, int val) {
        if (left == right) {
            seg[idx] = val;
            return;
        }
        
        int mid = (left + right) / 2;
        if (pos <= mid) {
            update(2 * idx, left, mid, pos, val);
        } else {
            update(2 * idx + 1, mid + 1, right, pos, val);
        }
        
        seg[idx] = combine(seg[2 * idx], seg[2 * idx + 1]);
    }
};

// Segment Tree with Lazy Propagation
class STLP {
public:
    vector<long long> seg, lazy, nums;
    int n;
    
    STLP(vector<int>& arr) : nums(arr.begin(), arr.end()), n(arr.size()) {
        seg.assign(4 * n, 0);
        lazy.assign(4 * n, 0);
        build(1, 0, n - 1);
    }
    
    long long combine(long long left, long long right) {
        return left + right;
    }
    
    void build(int idx, int left, int right) {
        if (left == right) {
            seg[idx] = nums[left];
            return;
        }
        
        int mid = (left + right) / 2;
        build(2 * idx, left, mid);
        build(2 * idx + 1, mid + 1, right);
        seg[idx] = combine(seg[2 * idx], seg[2 * idx + 1]);
    }
    
    void propagate(int idx, int left, int right) {
        if (lazy[idx] != 0) {
            seg[idx] += (right - left + 1) * lazy[idx];
            
            if (left != right) {
                lazy[2 * idx] += lazy[idx];
                lazy[2 * idx + 1] += lazy[idx];
            }
            
            lazy[idx] = 0;
        }
    }
    
    long long query(int idx, int left, int right, int ql, int qr) {
        propagate(idx, left, right);
        
        if (ql > right || qr < left) return 0;
        if (ql <= left && right <= qr) return seg[idx];
        
        int mid = (left + right) / 2;
        return combine(
            query(2 * idx, left, mid, ql, qr),
            query(2 * idx + 1, mid + 1, right, ql, qr)
        );
    }
    
    void update(int idx, int left, int right, int ql, int qr, int val) {
        propagate(idx, left, right);
        
        if (ql > right || qr < left) return;
        
        if (ql <= left && right <= qr) {
            lazy[idx] += val;
            propagate(idx, left, right);
            return;
        }
        
        int mid = (left + right) / 2;
        update(2 * idx, left, mid, ql, qr, val);
        update(2 * idx + 1, mid + 1, right, ql, qr, val);
        
        propagate(2 * idx, left, mid);
        propagate(2 * idx + 1, mid + 1, right);
        seg[idx] = combine(seg[2 * idx], seg[2 * idx + 1]);
    }
    
    long long rangeQuery(int left, int right) {
        return query(1, 0, n - 1, left, right);
    }
    
    void rangeUpdate(int left, int right, int val) {
        update(1, 0, n - 1, left, right, val);
    }
};

// Graph Algorithms
void dfs(vector<vector<int>>& graph, int node, vector<bool>& visited, vector<int>& res) {
    visited[node] = true;
    res.push_back(node);
    
    for (int nei : graph[node]) {
        if (!visited[nei]) {
            dfs(graph, nei, visited, res);
        }
    }
}

vector<int> bfsGraph(int root, vector<vector<int>>& graph) {
    vector<int> res;
    queue<int> q;
    set<int> seen;
    
    q.push(root);
    seen.insert(root);
    
    while (!q.empty()) {
        int node = q.front();
        q.pop();
        res.push_back(node);
        
        for (int nei : graph[node]) {
            if (seen.find(nei) == seen.end()) {
                q.push(nei);
                seen.insert(nei);
            }
        }
    }
    
    return res;
}

vector<int> topoSort(vector<pair<int, int>>& edges, int n) {
    vector<vector<int>> graph(n);
    vector<int> indegree(n, 0);
    
    for (auto& edge : edges) {
        graph[edge.first].push_back(edge.second);
        indegree[edge.second]++;
    }
    
    vector<int> topo;
    queue<int> q;
    
    for (int i = 0; i < n; i++) {
        if (indegree[i] == 0) {
            q.push(i);
        }
    }
    
    while (!q.empty()) {
        int node = q.front();
        q.pop();
        topo.push_back(node);
        
        for (int nei : graph[node]) {
            indegree[nei]--;
            if (indegree[nei] == 0) {
                q.push(nei);
            }
        }
    }
    
    if (topo.size() != n) return {};
    return topo;
}

bool isBipartite(vector<vector<int>>& graph, int n) {
    vector<int> color(n, -1);
    
    for (int start = 0; start < n; start++) {
        if (color[start] == -1) {
            queue<int> q;
            q.push(start);
            color[start] = 1;
            
            while (!q.empty()) {
                int node = q.front();
                q.pop();
                
                for (int nei : graph[node]) {
                    if (color[nei] == -1) {
                        color[nei] = 1 - color[node];
                        q.push(nei);
                    } else if (color[nei] == color[node]) {
                        return false;
                    }
                }
            }
        }
    }
    
    return true;
}

vector<int> dijkstra(int source, vector<vector<pair<int, int>>>& graph, int n) {
    vector<int> dist(n, INT_MAX);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    
    dist[source] = 0;
    pq.push({0, source});
    
    while (!pq.empty()) {
        int d = pq.top().first;
        int u = pq.top().second;
        pq.pop();
        
        if (d > dist[u]) continue;
        
        for (auto& edge : graph[u]) {
            int v = edge.first;
            int w = edge.second;
            
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                pq.push({dist[v], v});
            }
        }
    }
    
    return dist;
}

vector<int> bellmanFord(int source, vector<vector<pair<int, int>>>& graph, int n) {
    vector<int> dist(n, INT_MAX);
    dist[source] = 0;
    
    for (int i = 0; i < n - 1; i++) {
        for (int u = 0; u < n; u++) {
            if (dist[u] != INT_MAX) {
                for (auto& edge : graph[u]) {
                    int v = edge.first;
                    int w = edge.second;
                    if (dist[u] + w < dist[v]) {
                        dist[v] = dist[u] + w;
                    }
                }
            }
        }
    }
    
    for (int u = 0; u < n; u++) {
        if (dist[u] != INT_MAX) {
            for (auto& edge : graph[u]) {
                int v = edge.first;
                int w = edge.second;
                if (dist[u] + w < dist[v]) {
                    return {};
                }
            }
        }
    }
    
    return dist;
}

vector<vector<int>> floydWarshall(vector<tuple<int, int, int>>& edges, int n) {
    vector<vector<int>> dist(n, vector<int>(n, INT_MAX));
    
    for (int i = 0; i < n; i++) {
        dist[i][i] = 0;
    }
    
    for (auto& edge : edges) {
        int u = get<0>(edge);
        int v = get<1>(edge);
        int w = get<2>(edge);
        dist[u][v] = min(dist[u][v], w);
    }
    
    for (int k = 0; k < n; k++) {
        for (int u = 0; u < n; u++) {
            for (int v = 0; v < n; v++) {
                if (dist[u][k] != INT_MAX && dist[k][v] != INT_MAX) {
                    dist[u][v] = min(dist[u][v], dist[u][k] + dist[k][v]);
                }
            }
        }
    }
    
    for (int i = 0; i < n; i++) {
        if (dist[i][i] < 0) {
            return {};
        }
    }
    
    return dist;
}

// Disjoint Set Union
class DSU {
public:
    vector<int> parent, rank, size;
    int components;
    
    DSU(int n) : components(n) {
        parent.resize(n);
        rank.assign(n, 0);
        size.assign(n, 1);
        iota(parent.begin(), parent.end(), 0);
    }
    
    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }
    
    bool unite(int x, int y) {
        int px = find(x), py = find(y);
        if (px == py) return false;
        
        if (rank[px] < rank[py]) swap(px, py);
        
        parent[py] = px;
        if (rank[px] == rank[py]) rank[px]++;
        
        components--;
        return true;
    }
    
    bool uniteBySize(int x, int y) {
        int px = find(x), py = find(y);
        if (px == py) return false;
        
        if (size[px] < size[py]) swap(px, py);
        
        parent[py] = px;
        size[px] += size[py];
        
        components--;
        return true;
    }
    
    int getSize(int x) {
        return size[find(x)];
    }
    
    bool connected(int x, int y) {
        return find(x) == find(y);
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    return 0;
}